### PR TITLE
Table Component

### PR DESCRIPTION
**Pull request by MelSumner**
### :pushpin: Summary

If merged, this PR will add a table component to the design system.

Status, 11/23: ready to review

### :hammer_and_wrench: Detailed description

👀 Preview docs for this PR: https://hds-components-git-melsumner-table-hashicorp.vercel.app/components/table 

### :camera_flash: Screenshots

CleanShot 2022-11-08 at 12 28 15

***

### 👀 How to review

Reviewer's checklist:

- [ ] +1 Percy if applicable
- [ ] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

### CODE REVIEW

**Comment by MelSumner** [comment]
Ok so after some discussion we're going to take a different direction, so for now i am going to close this spike

File: packages/components/app/styles/components/table.scss [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    &.text-right { // for num support
+      text-align: right;
+    }
+    a { // might refactor this to a button TBH
+      align-items: center;
+      border: 3px solid transparent;
+      color: var(--token-color-foreground-strong);
+      display: flex;
+      padding: 8px 16px;
```
Comment (KristinLBradley) [inline packages/components/app/styles/components/table.scss:38]: Suggestion/Nit: Would you want to create a sass variable or something to avoid repeating these padding values?

Comment (MelSumner) [reply packages/components/app/styles/components/table.scss:38]: I think once they are settled then that seems reasonable, my first goal was to get them into a shape that mostly resembled what was in the figma file (which is still changing, so I'll probably wait to make additional style changes for right now) 👍

Comment (didoo) [reply packages/components/app/styles/components/table.scss:38]: Once the HTML/CSS code is finalized/settled, we can decide if (and which) compont-specific design tokens to introduce (similar to what we did for the Form controls, that paid off beyond expectations).

**Review by didoo — COMMENTED** [review]
I have looked at the proposed approach in general, not at single lines of code, as requested.
Here are some considerations.

From what I see, this approach deviates quite a bit from our current patterns, and I would like to see explained more in details why this choice, and what are the advantages of it (and how to mitigate the downsides).

Some examples of where I see it differ:
- we're not using the "contextual components" pattern for breaking down the component in sub-components; this patter is used in most of our (more structured) components, and has revealed itself as working very well for us and has been much appreciated by the developers
- developers need to express intent using class names instead of props/arguments (eg. to have a right-aligned `<td>` I would need to apply a class, instead of using an `@align` argument)
- we're manipulating data, instead of being agnostic and providing almost "pure" UI components (for all the other components we have delegated this responsibility to the consumers)
- we are not exposing sub-components as standalone/reusable components (there's no such thing as `<Hds::Table::Td>` at the moment) while we know this possibility has been much appreciated from the consumers/developers
- we're using descendants instead of class names for styling elements (this may be temporary though?) 

Also, one thing to consider that may be an issue: the current solution adds the query params to the URL, which I think we should not do it for these reasons:
- it's not responsibility of the component, we should provide hooks for the consumers if they want to do it (in case they need hooks, which I doubt if they're also in control of the data)
- it may interfere with other things in the page that may want to update the URL too, so we should be more agnostic and delegate this responsibility to the consumers
- even if we decide to do it, we need to make sure the query params don't conflict/interfere with other potential query params in the URL so we should prefix the query params with `hds-` to be 100% sure we're not overwriting other query params
- and even if we prefix these, we may also need to postfix the query params so they are unique (what if there are two HDS Table components in the same page, both sortable?)
so in conclusion, there are strong indication we should delegate this responsibility to the consumers.

As I said, more of a general feedback on the technical approach than a code review. Happy to discuss these points in detail in our meeting later today.

File: packages/components/addon/components/hds/table/index.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+/**
+ * @class HdsTable
+ */
+export default class HdsTableIndexComponent extends Component {
+  /**
+   * Sets the data source that gets yielded by the `:body` named block.
+   * @argument model
+   * @type {array}
+   */
```
Comment (jgwhite) [inline packages/components/addon/components/hds/table/index.js:11]: Could we add a `@key` argument that gets plumbed into the inner `{{#each}}`?

**Comment by jgwhite** [comment]
I’ve been using the cloud-ui table component quite a bit recently, and I’ve found it to be straightforward and fuss-free. From that perspective, I’m happy to see we’re sticking with the overall API.

I find myself on the fence regarding query params. I entirely agree with @didoo’s arguments, but I think it’s also true that it would be a useful default. I suppose one question to ask is “how complicated would it be for the consuming app to wrap Hds::Table in order to layer in query-params-by-default for the headings?”

> we're not using the "contextual components" pattern for breaking down the component in sub-components

I agree with this sentiment, but it seems to me there’s nothing preventing us evolving this component over time to add contextual components. More concretely, I could see Hds::Table landing in this form and a later version adding yielded components like `<T.Cell @align="right">`.

**Comment by didoo** [comment]
> there’s nothing preventing us evolving this component over time to add contextual components. More concretely, I could see Hds::Table landing in this form and a later version adding yielded components like <T.Cell @align="right">.

@jgwhite we're already struggling a lot getting a single migration of the components (once they're released) in Cloud UI, I can not imagine having to do these migrations incrementally, the cost would be too much in term of time and resources.

File: packages/components/addon/components/hds/table/index.js (hunk 2 above)
Comment (MelSumner) [reply packages/components/addon/components/hds/table/index.js:11]: @jgwhite I think we could, let me try it out!

**Comment by jgwhite** [comment]
> @jgwhite we're already struggling a lot getting a single migration of the components (once they're released) in Cloud UI, I can not imagine having to do these migrations multiple times, incrementally: the cost would be too much in term of time and resources.

@didoo ah ok, that’s fair. So, in this case, once there are a load of hand-written `<td>`s in the wild, it’s likely to be hard to migrate them to `<T.Cell>` or similar. Good data point.

File: packages/components/addon/components/hds/table/index.hbs [hunk 3]
```diff
@@ -0,0 +1,42 @@
+<table class={{this.classNames}} ...attributes>
+  {{#if this.isSortableTable}}
+    <caption class="sr-only" aria-live="polite">{{@caption}} {{this.sortedMessageText}}</caption>
+  {{else}}
+    <caption class="sr-only">{{@caption}}</caption>
+  {{/if}}
```
Comment (KristinLBradley) [inline packages/components/addon/components/hds/table/index.hbs:6]: Question/concern: Is the caption required? If it isn't, the caption would render as an empty tag with no value.

Comment (MelSumner) [reply packages/components/addon/components/hds/table/index.hbs:6]: That's correct; it could render as an empty component. In the case of the sortable table, it renders as empty if there's no caption and then updates the message once the sort has happened.

Comment (KristinLBradley) [reply packages/components/addon/components/hds/table/index.hbs:6]: Ok thanks. I would prefer to not render an empty element but I guess if you need it there to display a message for the sortable table after it's sorted I understand.

File: packages/components/addon/components/hds/table/index.hbs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            >
+              {{column.label}}
+            </Hds::Table::ThSort>
+          {{else}}
+            <Hds::Table::Th>{{column.label}}</Hds::Table::Th>
+          {{/if}}
+        {{/each}}
+      </Hds::Table::Tr>
+    {{else}}
+      {{yield (hash Tr=(component "hds/table/tr") Th=(component "hds/table/th")) to="head"}}
```
Comment (MelSumner) [inline packages/components/addon/components/hds/table/index.hbs:26]: I don't actually think this works properly because you can't do something like 

```
<Hds::Table as |T|>
  <:head>
    <T.Tr>
      <T.Th>Header One</T.Th>
     </T.Tr>
   <:/head>
 </Hds::Table>
 ```
 
 but I'm also wondering why it's working when I manually invoke the longhand form of the component if I've told it to yield the shorthand. 🤔

File: packages/components/tests/dummy/app/templates/components/table.hbs [hunk 5]
```diff
@@ -0,0 +1,689 @@
...
+<section>
+  <h3 class="dummy-h3" id="component-api">
+    <a href="#component-api" class="dummy-link-section">§</a>
+    Component API
+  </h3>
+  <p class="dummy-paragraph" id="component-api-table">
+    The
+    <code class="dummy-code">Table</code>
+    component itself is where most of options will be applied. Use of the component API (below) will automatically
+    render sortable columns. When the
```
[the remark this answers is not in the document]
Comment (alex-ju) [reply packages/components/tests/dummy/app/templates/components/table.hbs:32]: Even after merging the two sentences it still doesn't read very well – I'm sure you'll be able to reword this for clarity.

**Review by alex-ju — COMMENTED** [review]
Looks good overall – I agree with most of Cristiano's comments above, only spotted a few minor things aside from those.

File: packages/components/app/styles/components/table.scss [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+$hds-table-cell-padding-medium: 12px 16px;
+$hds-table-cell-padding-short: 4px 16px;
+$hds-table-cell-padding-tall: 20px 16px;
+
+.hds-table,
+.hds-table-sortable {
+  width: 100%;
+  font-size: var(--token-typography-body-200-font-size);
+  font-family: var(--token-typography-font-stack-text);
+  line-height: var(--token-typography-body-200-line-height);
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/app/styles/components/table.scss:18]: I see. What's our reason for this? I've always avoided inline styles as much as possible, so I'm curious.

File: packages/components/addon/components/hds/table/index.hbs [hunk 7]
```diff
@@ -0,0 +1,42 @@
+<table class={{this.classNames}} ...attributes>
+  {{#if this.isSortableTable}}
+    <caption class="sr-only" aria-live="polite">{{@caption}} {{this.sortedMessageText}}</caption>
+  {{else}}
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/table/index.hbs]: Added. 👍

File: packages/components/app/styles/components/table.scss [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+  border: $hds-table-border-width solid $hds-table-border-color;
+  border-radius: $hds-table-border-radius;
+  border-spacing: 0;
+}
+
+// TABLE HEADER
+.hds-table__thead {
+  .hds-table__tr {
+    color: var(--token-color-foreground-strong);
+    background-color: var(--token-color-surface-strong);
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/app/styles/components/table.scss:29]: Ah good catch! I'm going to take them off the th instead though, since they're on the tr in the tbody and it's also where folks generally expect to find these things (esp. re: striping)

File: packages/components/app/styles/components/table.scss [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+    .hds-table__th,
+    .hds-table__th-sort {
+      height: 48px;
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/app/styles/components/table.scss:33]: 1. will do!
2. good catch, fixed!

File: packages/components/app/styles/components/table.scss [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+          outline: 0;
+          box-shadow: inset 0 0 0 1px var(--token-color-focus-action-internal), 0 0 0 3px var(--token-color-focus-action-external);
+        }
+
+        &:active {
+          color: var(--token-color-foreground-strong);
+          background-color: var(--token-color-palette-neutral-300);
+        }
+
+        svg { // in case they don't use flight icons (yet)
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/app/styles/components/table.scss:86]: I'll change it, it can act as a forcing function (it'll be wrong unless a flight-icon is used).

File: packages/components/app/styles/components/table.scss [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+          margin-left: 8px;
+          color: var(--token-color-foreground-action);
+          vertical-align: middle; // we can't use display flex on the button, so we have to align the icon itself instead.
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/app/styles/components/table.scss:89]: this should be fixed with the update I'm pushing to support multi-line text 👍

File: packages/components/app/styles/components/table.scss (hunk 6 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/app/styles/components/table.scss:19]: These are not inline styles, these are inline class names (applied to HTML elements).

File: packages/components/app/styles/components/table.scss (hunk 10 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/app/styles/components/table.scss]: I see you had an issue with the `flight` icon name (usually we use local `hds` names).
I have opened a PR to update the `stylelint` configuration to allow also `flight` prefixes: https://github.com/hashicorp/design-system/pull/696

(once merged an your tests are passing, please remove the comment you added)

File: packages/components/app/styles/components/table.scss (hunk 11 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/app/styles/components/table.scss]: @MelSumner notice: with the new layout/elements organization, you have to set the height of the `button` to `100%` or this will happen:
screenshot_2046

File: packages/components/addon/components/hds/table/index.hbs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        {{#each this.columns as |column|}}
+          {{#if column.isSortableColumn}}
+            <Hds::Table::ThSort
+              @sortKey={{column.key}}
+              @sortBy={{this.sortBy}}
+              @sortOrder={{this.sortOrder}}
+              @onClick={{this.setSortBy}}
+            >
+              {{column.label}}
+            </Hds::Table::ThSort>
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/table/index.hbs:19]: ok changed this. 👍

File: packages/components/tests/dummy/app/templates/components/table.hbs [hunk 13]
```diff
@@ -0,0 +1,689 @@
...
+  />
+  {{! template-lint-enable no-unbalanced-curlies }}
+  {{! prettier-ignore-end }}
+  <h4 class="dummy-h4">Simple Table with model defined</h4>
+  <p class="dummy-paragraph">In the simplest invocation of the table component, you would define the model and insert
+    your own content into the
+    <code class="dummy-code">:head</code>
+    and
+    <code class="dummy-code">:body</code>
+    blocks. Here is an example of such an invocation in a template:</p>
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/tests/dummy/app/templates/components/table.hbs:158]: come back to this after thinking about it some more (I thought I already gave this information)

File: packages/components/tests/dummy/app/templates/components/table.hbs [hunk 14]
```diff
@@ -0,0 +1,689 @@
...
+  <p class="dummy-paragraph" id="component-api-table">
+    The
+    <code class="dummy-code">Table</code>
+    component itself is where most of options will be applied. Use of the component API (below) will automatically
+    render sortable columns. When the
+    <code class="dummy-code">@columns</code>
+    property is invoked with a value.
+  </p>
+  <p class="dummy-paragraph">
+    Additionally, there are child components that can also be used to provide custom options.
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/tests/dummy/app/templates/components/table.hbs:35]: for now it's okay

File: packages/components/tests/integration/components/hds/table/index-test.js [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+  ]);
+  context.set('columns', [
+    { key: 'artist', label: 'Artist' },
+    { key: 'album', label: 'Album' },
+    { key: 'year', label: 'Year' },
+  ]);
+  context.set('sortBy', 'artist');
+  context.set('sortOrder', 'asc');
+};
+const renderSortableTable = async () => {
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/tests/integration/components/hds/table/index-test.js:39]: @didoo will PR

File: packages/components/addon/components/hds/table/index.hbs [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+          {{/if}}
+        {{/each}}
+      </Hds::Table::Tr>
+    {{else}}
+      {{yield (hash Tr=(component "hds/table/tr") Th=(component "hds/table/th")) to="head"}}
+    {{/if}}
+  </thead>
+  <tbody class="hds-table__tbody">
+    {{#if this.isSortableTable}}
+      {{#each (sort-by this.getSortCriteria @model) as |data|}}
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/table/index.hbs:30]: we're going to come back to this one because we need to think more about the td situation.

File: packages/components/tests/integration/components/hds/table/index-test.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/tests/integration/components/hds/table/index-test.js:39]: @melanie here are two alternative PRs about this particular code:
- https://github.com/hashicorp/design-system/pull/700
- https://github.com/hashicorp/design-system/pull/701

Pick one that you think is better/more clear for you.

**Review by alex-ju — APPROVED** [review]
:shipit:
